### PR TITLE
Add minutes from 2017-03-14 HTTP meeting

### PR DESCRIPTION
**Pull request by seabaylea**
Minutes from the 2017-03-14 HTTP meeting

@swift-server/stakeholders please review.

### CODE REVIEW

File: meetings/http/2017-03-14/minutes.md [hunk 1]
```diff
@@ -1,24 +1,65 @@
...
+* Michael Papp
+* Thomas Catterall
+
+### Minutes:
+
+_Use of value types vs. reference types for request and response_ 
+* General consensus is in favour of using a value types for request and response.
+  * The argument that the should be references as they are passed around and mutated really don’t hold
+  * It should be noted that the reason why there is still NSURLRequest/Response (reference types) was because a subclass exists, rather than the intend that they should be reference types.
+  * The use of value types will help with reference cycles in closures (like middlewares) which you otherwise need to use weak self for.
```
Comment (helje5) [inline meetings/http/2017-03-14/minutes.md]: Is there an example for this? A request doesn't refer to middleware, and the middleware gets passed in the req/res as arguments, why would there ever be a cycle?

File: meetings/http/2017-03-14/minutes.md [hunk 2]
```diff
@@ -1,24 +1,65 @@
...
+  * Should also help with reducing some of the ARC overhead
+* ARC overhead won’t necessarily be removed if the struct/value types has references inside it - and the body data is really a stream which has to be a reference type
+  * In Foundation, request and response is just the headers, the body data is a side stream so they can be separated
```
Comment (helje5) [inline meetings/http/2017-03-14/minutes.md:40]: The (decoded) body stream can be separated from the headers, but the underlying protocol stream *cannot* be separated from the headers. In HTTP/1.1 headers can be sent *after* body data. (E.g. check out https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer.)

Or in other words: the header set is only complete after the whole HTTP message has been processed.

You could craft a different API for headers coming in later, but that seems a little weird to me and probably leads to a 'fix later' situation in which trailing headers are not properly implemented. Unless you do the separate API, it seems the incoming message *has* to be a reference type.

File: meetings/http/2017-03-14/minutes.md [hunk 3]
```diff
@@ -1,24 +1,65 @@
...
+  * This is done URLSession (as with downloadTask) into a file or a stream (and the same would be true for responses).
+  * Can be thought of like sequence, which is a value type for which you can get an iterator which is a reference type
+  * Separate experience from Johannes is creating a framework which has request/response separate from the body stream.
+    * Request and response are value types
+    * Handler provides a BodyProcessor which is called when chunks of data become available
+    * This is especially important when the client is using Chunked transfer encoding, which some frameworks currently don’t attempt to handle.
+* General consensus in separating the request and response headers (metadata) away from the body, which can be handled separately either by a handler API or a closure.
+* How do we handle HTTP trailers if we separate headers and body stream?
+  * Trailers are in the body stream and arrive after the headers are complete, but add to the headers
+  * These aren’t well specified, and probably need to be handled by whatever is processing the stream anyway.
```
Comment (helje5) [inline meetings/http/2017-03-14/minutes.md:50]: What about them isn't well specified? E.g. hashing is an important and valid use case for trailing headers.

File: meetings/http/2017-03-14/minutes.md [hunk 4]
```diff
@@ -1,24 +1,65 @@
...
+
+### Minutes:
+
+_Use of value types vs. reference types for request and response_ 
+* General consensus is in favour of using a value types for request and response.
+  * The argument that the should be references as they are passed around and mutated really don’t hold
+  * It should be noted that the reason why there is still NSURLRequest/Response (reference types) was because a subclass exists, rather than the intend that they should be reference types.
+  * The use of value types will help with reference cycles in closures (like middlewares) which you otherwise need to use weak self for.
+  * Should also help with reducing some of the ARC overhead
+* ARC overhead won’t necessarily be removed if the struct/value types has references inside it - and the body data is really a stream which has to be a reference type
```
Comment (helje5) [inline meetings/http/2017-03-14/minutes.md]: Well, for GET, which is a quite common request, no reference would be necessary ;-> I see the bigger issue with all the copying of the header block (which is not _huge_ but still significant) as it goes through middleware. Presumably a single ARC is cheaper. IMO to come to a proper conclusion on that a performance test is warranted instead of claims('should help etc').

File: meetings/http/2017-03-14/minutes.md (hunk 2 above)
Comment (ghost) [reply meetings/http/2017-03-14/minutes.md:40]: With Chunked Transfer Encoding, there can also be Trailer headers per chunk (e.g. chunk CRC that relates only to the body of that single chunk), so it's not that easy to define what the complete header set should look like.

File: meetings/http/2017-03-14/minutes.md [hunk 5]
```diff
@@ -1,24 +1,65 @@
...
+* Nasser Ebrahim
+* Steve Algernon
+* Michael Papp
+* Thomas Catterall
+
+### Minutes:
+
+_Use of value types vs. reference types for request and response_ 
+* General consensus is in favour of using a value types for request and response.
+  * The argument that the should be references as they are passed around and mutated really don’t hold
```
Comment (Baza207) [inline meetings/http/2017-03-14/minutes.md]: I hate to be that guy, but shouldn't this be:
> The argument that _**they**_ should...

🙂

File: meetings/http/2017-03-14/minutes.md (hunk 1 above)
Comment (shmuelk) [reply meetings/http/2017-03-14/minutes.md]: Depending on how middleware are invoked, middleware may need to "insert hooks" to get invoked in later points in the lifecycle of a request. Examples that come to mind are middleware like session support (needs to be able to store the session and add to the reponse (probably a session id cookie) and compression.

Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: Is that a response to my question or a general comment? If so: What do you mean by "insert hooks"? Where does the cycle happen?

Neither session middleware nor compression seem to be candidates for retain cycles. (and compression shouldn't be a middleware in the first place, but a stream)

File: meetings/http/2017-03-14/minutes.md [hunk 6]
```diff
@@ -1,24 +1,65 @@
...
+* Marcin Kliks
+* Nasser Ebrahim
+* Steve Algernon
+* Michael Papp
+* Thomas Catterall
+
+### Minutes:
+
+_Use of value types vs. reference types for request and response_ 
+* General consensus is in favour of using a value types for request and response.
```
Comment (loganwright) [inline meetings/http/2017-03-14/minutes.md:34]: I feel like this is a knee-jerk reaction to swift making things like structs easier to build, however in practice, value types (in req/resp models) cause a lot of problems, create more confusing middleware, and introduce more bugs than reference types.

I know value types and structs are the new hotness, but in a server side context they just don't add enough value.

File: meetings/http/2017-03-14/minutes.md (hunk 1 above)
Comment (loganwright) [inline meetings/http/2017-03-14/minutes.md]: I don't think this is a valid argument, middleware shouldn't be retaining a request in this way because then it would be holding a single request in perpetuity. Requests should be passed in as arguments.

The notion of weak self in terms of other components of an application are problematic regardless of the value/reference semantics of request/response.

File: meetings/http/2017-03-14/minutes.md [hunk 7]
```diff
@@ -1,24 +1,65 @@
...
+  * The argument that the should be references as they are passed around and mutated really don’t hold
+  * It should be noted that the reason why there is still NSURLRequest/Response (reference types) was because a subclass exists, rather than the intend that they should be reference types.
+  * The use of value types will help with reference cycles in closures (like middlewares) which you otherwise need to use weak self for.
+  * Should also help with reducing some of the ARC overhead
```
Comment (loganwright) [inline meetings/http/2017-03-14/minutes.md]: In our experience, value types increased copy overhead leading to major performance declines. ARC is nominal and considerably more performant. I don't think this argument is relevant.

File: meetings/http/2017-03-14/minutes.md [hunk 8]
```diff
@@ -1,24 +1,65 @@
...
+* ARC overhead won’t necessarily be removed if the struct/value types has references inside it - and the body data is really a stream which has to be a reference type
+  * In Foundation, request and response is just the headers, the body data is a side stream so they can be separated
+  * This is done URLSession (as with downloadTask) into a file or a stream (and the same would be true for responses).
+  * Can be thought of like sequence, which is a value type for which you can get an iterator which is a reference type
+  * Separate experience from Johannes is creating a framework which has request/response separate from the body stream.
+    * Request and response are value types
+    * Handler provides a BodyProcessor which is called when chunks of data become available
+    * This is especially important when the client is using Chunked transfer encoding, which some frameworks currently don’t attempt to handle.
```
Comment (LoganWright) [inline meetings/http/2017-03-14/minutes.md]: Chunked transfer encoding according to RFC I believe says it should be collected by the server into the model before passing it on to end consumer.

More flexibility here would probably be nice though.

File: meetings/http/2017-03-14/minutes.md [hunk 9]
```diff
@@ -1,24 +1,65 @@
...
+* General consensus in separating the request and response headers (metadata) away from the body, which can be handled separately either by a handler API or a closure.
```
Comment (loganwright) [inline meetings/http/2017-03-14/minutes.md:47]: This seems like it could be useful as long as the 95% use case isn't particularly difficult.

**Review by loganwright — COMMENTED** [review]
I really feel that the issue of reference semantics and making request/response value type is not the correct path for the group. Having worked with both extensively in a server side scenario, the number of problems created by value types greatly outweigh any minor benefit we might get for participating in the trendy new value capabilities.

This feels like a classic case of just because we can do something doesn't mean we should. I'd prefer to be able to integrate as natively as possible, but given this architecture, Vapor has no choice but to wrap or convert which is tedious and slower.

If we really can't let go of the value trend, I'd like to propose that we avoid concrete models as much as possible and instead opt for protocols in all serialization and parser layers. This would perhaps be even better in long run because it would allow consumers to provide their own models.

File: meetings/http/2017-03-14/minutes.md (hunk 9 above)
Comment (swizzlr) [reply meetings/http/2017-03-14/minutes.md:47]: @LoganWright Seems like it wouldn't be terribly awkward for Frameworks to provide their own type that wraps both of these, or even (better?) consume the entire request stream first, then provide to client code.

File: meetings/http/2017-03-14/minutes.md (hunk 6 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md:34]: @LoganWright Keep in mind that this is supposed to be a low level library. I would assume that the request thing this produces is almost always going to be wrapped by a higher level library like Kitura etc to provide the framework specific features. This effort is more the Node.js with Kitura being something sitting on top, like Express.js.

File: meetings/http/2017-03-14/minutes.md (hunk 7 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: Maybe someone can correct me, but the big issue with ARC is that it needs to be thread safe even if no threads are involved, hence passing references around is not just a simple `rc+=1`. Not as bad as the Python GIL, but kinda like a LIL ;->

File: meetings/http/2017-03-14/minutes.md (hunk 1 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: Just a sidenote: I think the original bullet point is a little misleading. I don't think that this was about `middleware` itself, but middleware attaching a closure to a request (say a persist-handler for a session which is setup when the session middleware runs, but invoked later, once the request processing ends).

File: meetings/http/2017-03-14/minutes.md (hunk 6 above)
Comment (loganwright) [reply meetings/http/2017-03-14/minutes.md:34]: @helje5 if that's the case, then I'd much prefer to just accept protocols with the required fields in an initializer of some sort for parsing and properties of some sort for serialization as there's no need for models.

When there's discussion of included middleware etc. then this immediately jumps up a level and reference semantics matter. In these situations, Value semantics are more difficult to work with and cause more bugs.

File: meetings/http/2017-03-14/minutes.md (hunk 7 above)
Comment (loganwright) [reply meetings/http/2017-03-14/minutes.md]: Either way, that's not something that we're responsible for, arc is handled by a lower level system and I don't think adds relevant points to this discussion. Copy overhead significantly outweighs any overhead that might be incurred by arc under the hood.

File: meetings/http/2017-03-14/minutes.md (hunk 9 above)
Comment (loganwright) [reply meetings/http/2017-03-14/minutes.md:47]: @swizzlr falls back to a point I was making earlier. If the point of all this is that nobody's using the types, why are we bothering with types at all where protocols and generics would be fully sufficient. Maybe that's a better solution.

File: meetings/http/2017-03-14/minutes.md (hunk 5 above)
Comment (gwynne) [reply meetings/http/2017-03-14/minutes.md]: I would use the word "these" in that context. Also "doesn't" instead of "don't", while we're talking grammar 😉

File: meetings/http/2017-03-14/minutes.md (hunk 2 above)
Comment (gwynne) [reply meetings/http/2017-03-14/minutes.md:40]: Does anyone actually use chunk trailers in practice? How large a concern is this?

File: meetings/http/2017-03-14/minutes.md (hunk 9 above)
Comment (gwynne) [reply meetings/http/2017-03-14/minutes.md:47]: Agreed; protocols seem to make more sense here, especially in the presence of common middleware. The only concern I have is whether Swift's generics system as it presently stands can handle the appropriate constraints.

File: meetings/http/2017-03-14/minutes.md (hunk 3 above)
Comment (gwynne) [reply meetings/http/2017-03-14/minutes.md:50]: I think it's more accurate to say here that their _uses_ are not strongly defined. The spec does not provide for any common use case, meaning no single implementation of any kind is guaranteed to support a particular usage or even interpret trailing headers at all. (Speaking from my own experience, last time I wrote a chunked encoding decoder, I processed trailing headers and then discarded them because finding a reasonable place for them in the data model added great complexity for no tangible benefit.)

File: meetings/http/2017-03-14/minutes.md [hunk 10]
```diff
@@ -1,24 +1,65 @@
...
+* How do we handle HTTP trailers if we separate headers and body stream?
+  * Trailers are in the body stream and arrive after the headers are complete, but add to the headers
+  * These aren’t well specified, and probably need to be handled by whatever is processing the stream anyway.
+
+_Use of Foundation based types for request and response_ 
+* It’s accepted that the Foundation types don’t match the requirements for general request and response types today as their designed for a specific set of use cases.
```
Comment (gwynne) [inline meetings/http/2017-03-14/minutes.md]: Grammar nit: "they're" or "they are" instead of "their".

File: meetings/http/2017-03-14/minutes.md (hunk 6 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md:34]: @LoganWright I'm a little split on the protocols, but yes, I think it may be worthwhile to use such. I see this effort more in providing a standard library which gets you started quickly, like the `http` modules in Python. A decent, working implementation, maintained by a 'standards' body. At the same time I'm pretty sure that most higher level frameworks are still going to use their own stack for performance reasons (it is reasonably easy to use http_parser directly, for most framework authors there is simply no need for a Swift wrapper). Even protocols make things slower than for example using generics for parsing, as we figured out with the http_parser performance porting tests.

File: meetings/http/2017-03-14/minutes.md (hunk 2 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md:40]: Well, HTTP/1.1 is not some arbitrary thing people came up with on a weekend. Chunk trailers wouldn't be in there if it wouldn't have been important for someone relevant. I don't see it as a big issue either, it just means that there needs to be an API for this. Or not :-) Who uses TE, pipelining, multiple ETags etc ... I would recommend to stick to the spec and just do what it says.

File: meetings/http/2017-03-14/minutes.md (hunk 7 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: Sorry, but that just doesn't make any sense. It is like saying that the GIL doesn't matter to Python applications and is just an implementation detail. Locking overhead is a BIG concern on any current platform and picking a sync free data structure over one requiring sync may have significant advantages even if the overhead may be a little higher in a single thread performance test.

Comment (michael-yuji) [reply meetings/http/2017-03-14/minutes.md]: The ARC overhead is actually less likely to overcome the copy overhead.  I don't think a response/request will processed simultaneously on different threads very often so it should be pretty thread safe. What I concern about is the overhead in memory allocation for swift object members.

Personally I prefer a value type but reference semantic design. The data portion, and  anything will send/recv to/from network, including http headers, should pass by reference, any other members of the object should be small and pass by value instead, to avoid memory allocation overhead. 

The ARC shouldn't be an issue, even if it is an issue, it is not our problem but the compiler team. Even if the ARC do introduce a huge overhead, we can still eliminate the overhead by using `Unmanaged`, or, simply implement things in the C way.

Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: This is all nice and right, but you are mixing up ARC with `UnsafePointer<struct>`. I'm 100% in for using a zero copy architecture, but in Swift that would require the latter (no variable sized structs or objects). ARC has a non-negible synchronisation overhead, it is not the kind of issue the compiler can just fix for you. In contrast (depending on the actual ABI) the compiler *can* reduce copying of structs (if it's all 'let's, there is no reason to copy anything, while the compiler can't know which threads have a retain on an object).

Please be careful with statements like that. There is way too much 'less likely' and 'shouldn't be' for my taste ;-)
To provide an actual example you can profile, check out this [http-c-vs-swift](https://github.com/helje5/http-c-vs-swift). ARC is a huge overhead in that, and the example doesn't even do a lot of referencing back-n-forth :-)

Using objects also has the disadvantage that they need be heap allocated (not sure whether that is strictly true in Swift :-), but I assume it will stay that way). Often times that can be yet another major performance issue which can be avoided with stack allocated structs. (e.g. very little of the data of the HTTP message head should/can be zero copied)

Finally your example refers to net map which is mainly concerned about the block I/O which relates to the design of the stream part of the request, *not* the HTTP message heads. Yes, we absolutely should not copy body parts of the HTTP stream (and I'm still a little more interested how the stream API will look like).

In summary I guess my opinion is that: 
- What the HTTP Swift library passes out is likely going to be wrapped, either a custom "IncomingRequest" object, or a sideline object similar to `WOContext`.
- I don't care too much whether the request is done as a struct or object, my gut feeling is that the struct will be more performant (for the low-level data holder, *not* for the object passed around at high levels). Just guesswork, needs to be tested.
- I think it his highly desirable that client and server share the same infrastructure
- It may be desirable to have a protocol for Request/Response which all Swift server frameworks can share.
- Unfortunately I'm pretty sure this is going to be just the 'default implementation', performance oriented frameworks will do their own thing anyways. Parsing HTTP headers has a lot of optimisation potential, particularly in Swift.

File: meetings/http/2017-03-14/minutes.md (hunk 9 above)
Comment (swizzlr) [reply meetings/http/2017-03-14/minutes.md:47]: Indeed, a framework ought to support (or reject) the notion of middlewares (or arbitrary threads modifying data at a later point in time).

A protocol would be wonderful.

File: meetings/http/2017-03-14/minutes.md (hunk 7 above)
Comment (tanner0101) [reply meetings/http/2017-03-14/minutes.md]: The discussion about ARC seems like a moot point since requests and responses are not passed between threads in a dispatch based web server.

Furthermore, focusing on the implementation details of the compiler more than the actual design and function of the API seems backward to me: 1: It doesn't matter how performant an API is if it's difficult to use/easy to use incorrectly 2: implementation details of a compiler should not be relied upon. 

If we take a step back: the client makes one unique request to the server and the server responds with one unique response. This is more indicative of a reference type than a value type. 

If you get a request passed into a method body, and you mutate a property, you'd expect it to modify the original request. It just doesn't make sense for a request to function like an integer or a string would. It's worth noting this claim is indeed based in experience: One of the early alpha versions of Vapor had value semantics on request and response and it created a TON of bugs and confusion from developers.

TL;DR: can we focus on the API usability instead of the compiler's implementation details?

Comment (shmuelk) [reply meetings/http/2017-03-14/minutes.md]: I think most  of folks on the list who are in favor of value types here, are of the belief that they should be passed around as inout's. I grant that it can be claimed that value types passed around as inout's kind of defeats the idea of immutability of value types. The Apple folks claim that it is better than reference types because it is clearer as to what is going to happen when you pass the type around.

File: meetings/http/2017-03-14/minutes.md (hunk 6 above)
Comment (d6u) [reply meetings/http/2017-03-14/minutes.md:34]: @LoganWright 

> however in practice, value types (in req/resp models) cause a lot of problems, create more confusing middleware, and introduce more bugs than reference types.

I might be asking too much, but curious on some examples?

File: meetings/http/2017-03-14/minutes.md (hunk 7 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: > The discussion about ARC seems like a moot point since requests and responses are not passed between threads in a dispatch based web server

It is not a moot point, I'm a little confused. You must be aware that the problem with ARC is that the sync overhead occurs whether or not threads are actually in use? The compiler simply doesn't know about that and hence has to apply the overhead *always*. Which is unlike struct copying in which the compiler actually has the ability to optimize a lot of copying away (there is no real reason to ever copy a let struct the way down the call hierarchy, etc). It is just a matter of fact.

Yes, the optimizer can try to reduce the actual ARC operations (presumably it does way to many of them now, partially to please ObjC), but the fact stands that passing an ARC reference is a synchronized operation and that a synchronized operation in todays platforms can be significantly more expensive than copying 10 or even 100 times the data.

If you *know* that a request is not passed between threads, an `UnsafePointer<struct>` may be the way to go. This just passes the reference and doesn't incur the ARC overhead (which again is not incrementing a counter, but *atomically* incrementing a counter). Or maybe Swift 4 gets a way to mark references as 'single-threaded', that would be an option too?

I agree with you that performance is not everything and that API usability is important too, but since we are talking about low level frameworks here, not about the thing the actual developer is going to see (as you are going to have nice-Kitura APIs on top for them), maybe it is worth doing a proper evaluation.
Or we are talking about a higher-level framework in the sense of the Python `http` module which probably means that most performance sensitive frameworks will avoid it, as the `http` module.

File: meetings/http/2017-03-14/minutes.md (hunk 1 above)
Comment (seabaylea) [reply meetings/http/2017-03-14/minutes.md]: The problematic case is that middlewares sometimes (often?) invoke closures, which opens the opportunity for a reference cycle between the closure and a captured self reference. We're seeing this increasingly frequently in user code.

File: meetings/http/2017-03-14/minutes.md (hunk 4 above)
Comment (seabaylea) [reply meetings/http/2017-03-14/minutes.md]: The approach would be to make use of `inout`, because ownership stays with the original request handler but middlewares need access to it.

I agree - post-prototyping we'll be able to validate the approach through benchmarking against a number of existing parses (that use reference types today).

File: meetings/http/2017-03-14/minutes.md (hunk 6 above)
Comment (seabaylea) [reply meetings/http/2017-03-14/minutes.md:34]: @LoganWright whilst there's discussion of middlewares, the intent is not for the HTTP APIs to define how middlewares would be implemented - that's very much for the frameworks to determine. There is however a desire for request/response to be "middleware friendly" as that is a major server-side use case.

File: meetings/http/2017-03-14/minutes.md (hunk 1 above)
Comment (helje5) [reply meetings/http/2017-03-14/minutes.md]: @seabaylea I don't disagree that closures w/ ARC are problematic, but the raw HTTP request discussed here doesn't have any references to user or middleware provided closures. In fact it likely doesn't have any object references, it just carries strings and maybe enums. How would that ever introduce a reference cycle?

Presumably your specific framework implementation of the request object doesn't only carry the HTTP information but also has extra storage in which middleware can attach stuff (e.g. closures to be run later). How you implement middleware in Kitura is your choice :-) Presumably the Kitura.IncomingRequest could be a struct holding the HTTPServer.Request plus all the extra fields you need, if that is what you want.
